File: backend/app/services/weather_service.py
import requests
import os
from datetime import datetime, timezone

API_KEY = os.getenv("OPENWEATHER_API_KEY")
BASE_URL = "https://api.openweathermap.org/data/2.5/weather"

def fetch_weather_by_city(city: str):
    try:
        params = {
            "q": city,
            "units": "metric",
            "appid": API_KEY
        }
        response = requests.get(BASE_URL, params=params)
        response.raise_for_status()
        data = response.json()
        return parse_weather_response(data)
    except requests.RequestException as e:
        return {"error": f"Weather API request failed: {str(e)}"}

def fetch_weather_by_coords(lat: float, lon: float):
    try:
        params = {
            "lat": lat,
            "lon": lon,
            "units": "metric",
            "appid": API_KEY
        }
        response = requests.get(BASE_URL, params=params)
        response.raise_for_status()
        data = response.json()
        return parse_weather_response(data)
    except requests.RequestException as e:
        return {"error": f"Weather API request failed: {str(e)}"}

def parse_weather_response(data):
    if "main" not in data or "weather" not in data:
        return {"error": "Invalid response from weather API"}
    
    return {
        "temperature": data["main"]["temp"],
        "condition": data["weather"][0]["main"],
        "humidity": data["main"]["humidity"],
        "windSpeed": data["wind"]["speed"],
        "location": data["name"],
        "timestamp": datetime.now(timezone.utc).isoformat()
    }

File: Closet Companion/Services/WeatherService.swift
//
//  WeatherService.swift
//  Closet Companion
//
//  Created by Ian Bailey on 2/18/25.
//

import Foundation
import CoreLocation
import Combine

class WeatherService: ObservableObject {
    // MARK: - Properties
    @Published var currentWeather: WeatherData = .defaultWeather
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let apiKey = "YOUR_API_KEY" // Replace with your actual API key
    private let baseURL = "https://api.openweathermap.org/data/2.5/weather"
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Public Methods
    
    /// Fetch weather for a specific location by city name
    func fetchWeather(for city: String) {
        isLoading = true
        errorMessage = nil
        
        guard let encodedCity = city.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
            self.errorMessage = "Invalid city name"
            self.isLoading = false
            return
        }
        
        let urlString = "\(baseURL)?q=\(encodedCity)&units=metric&appid=\(apiKey)"
        
        guard let url = URL(string: urlString) else {
            self.errorMessage = "Invalid URL"
            self.isLoading = false
            return
        }
        
        fetchWeatherData(from: url)
    }
    
    /// Fetch weather for a specific location by coordinates
    func fetchWeather(latitude: Double, longitude: Double) {
        isLoading = true
        errorMessage = nil
        
        let urlString = "\(baseURL)?lat=\(latitude)&lon=\(longitude)&units=metric&appid=\(apiKey)"
        
        guard let url = URL(string: urlString) else {
            self.errorMessage = "Invalid URL"
            self.isLoading = false
            return
        }
        
        fetchWeatherData(from: url)
    }
    
    /// Fetch weather for the user's current location
    func fetchWeatherForCurrentLocation(location: CLLocation) {
        fetchWeather(latitude: location.coordinate.latitude, longitude: location.coordinate.longitude)
    }
    
    // MARK: - Private Methods
    
    private func fetchWeatherData(from url: URL) {
        URLSession.shared.dataTaskPublisher(for: url)
            .map(\.data)
            .receive(on: DispatchQueue.main)
            .tryMap { data -> WeatherData in
                // Try to decode the data
                let decoder = JSONDecoder()
                
                // Parse OpenWeatherMap JSON response
                let response = try decoder.decode(OpenWeatherResponse.self, from: data)
                
                // Convert to our WeatherData model
                return WeatherData(
                    temperature: response.main.temp,
                    condition: response.weather.first?.main ?? "Unknown",
                    humidity: response.main.humidity,
                    windSpeed: response.wind.speed,
                    location: response.name,
                    timestamp: Date()
                )
            }
            .sink { completion in
                self.isLoading = false
                
                if case .failure(let error) = completion {
                    self.errorMessage = "Failed to fetch weather: \(error.localizedDescription)"
                }
            } receiveValue: { weatherData in
                self.currentWeather = weatherData
                self.errorMessage = nil
            }
            .store(in: &cancellables)
    }
}

// MARK: - API Response Structures

// These structures match the OpenWeatherMap API response format
private struct OpenWeatherResponse: Codable {
    let weather: [WeatherInfo]
    let main: MainInfo
    let wind: WindInfo
    let name: String
}

private struct WeatherInfo: Codable {
    let id: Int
    let main: String
    let description: String
    let icon: String
}

private struct MainInfo: Codable {
    let temp: Double
    let humidity: Int
}

private struct WindInfo: Codable {
    let speed: Double
}

